pkg/logger/options.go

```diff
@@ -0,0 +1,110 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package logger
+
+import (
+	"fmt"
+)
+
+const (
+	defaultJSONOutput  = false
+	defaultOutputLevel = InfoLevel
+	undefinedDaprID    = ""
+)
+
+var stringToDaprLogLevel = map[string]LogLevel{
```

Comment:
Related to comment about string types: keeping reverse maps should be avoided **if possible**.

Reply:
I have two implementation goals here:
1. Do not use any logrus related things outside `logger.go`
2. Make tests easier without logrus dependency (e.g. please see options_test.go)

If we use `logrus.ParseLevel` to parse `-log-level` option, then it means we rely on logrus to parse command flags.

Additionally, there will be a chance to switch to new logger due to the performance if production users want to log all requests between daprd and darpd and between daprd and user as a log, not as a trace.

In my previous team, I always tracked the requests between services and between customer to services as a log, not a trace. In many cases, it is really useful to resolve the incidents. If I were using Dapr in my production service, I would always track all requests for handling the incidents. 

You may disagree with tracking the request due to the performance. But, I believe that customers in productions will want to track all. So, there would be a chance to move from logrus to high-performance logger such as zap like the k8s, knative, and istio. (I know that you're concerning about the size of zap) In this case, this implementation doesn't require any changes outside `logger.go`.

`stringToDaprLogLevel` is to define our own log levels and parse cmd flags **without logrus dependency**. such mapping table or mapping function is the simplest way to check the validity of flags in options.go implementation.

Reply:
This implementation is already tightly coupled to logrus, and is highly dependent on it. 

It uses logrus structs and methods everywhere inside `logger`.
If we ever want to switch to something like `zap`, we would need to change code inside `logger.go`.

So the current design is built around logrus, which by the way, it totally fine with me because changing it would be easy.

The change I'm suggesting will let us keep both bullet points you mentioned above.

My suggestion is not to expose logrus structs to the caller, it means that we keep the abstractions we currently have, but expose them as string types to avoid declaring multiple maps for reverse lookups which hurt maintainability.

Reply:
> This implementation is already tightly coupled to logrus, and is highly dependent on it.
> It uses logrus structs and methods everywhere inside logger.

I already mentioned it in my goals. "Do not use any logrus related things outside logger.go" which means only logger.go uses logrus. 

> So the current design is built around logrus

No this is not true, the methods in `Logger` interface are all Dapr-specific, not logrus. I reviewed the multiple go logging libraries including zap. The logging levels they supported are same or similar. The other methods are dapr-specific. daprLogger struct includes the actual logrus implementation. so it is easy to switch to zap later if we need.

> expose them as string types to avoid declaring multiple maps for reverse lookups which hurt maintainability.

I will work on it. but it requires something else to check the validity of logging level given by user.

Reply:
> No this is not true, the methods in Logger interface are all Dapr-specific, not logrus. I reviewed the multiple go logging libraries including zap. The logging levels they supported are same or similar. The other methods are dapr-specific. daprLogger struct includes the actual logrus implementation. so it is easy to switch to zap later if we need.

The interface is Dapr specific and is aligned with most logger implementations, that's true. What I mean is that even in `logger.go`, the methods which use `logrus` can be abstracted out and moved to a sub-package to keep it cleaner. But as I mentioned, I don't think that's needed and the current implementation is fine, since there will always ever be a single logger implementation, and `logger.go` is lightweight.

> I will work on it. but it requires something else to check the validity of logging level given by user.

Thanks, that's really all I asked to be changed..

By the way, since the logger specific implementation occurs in `logger.go` and it is being called in the ApplyOptions function in `options.go`, (like this `v.SetOutputLevel(options.GetOutputLevel())`, the parsing of the log level can move to `logger.go` and an error will be returned from ApplyOptions. This actually more natural to me, because you give much more freedom to the actual logger implementation rather than having Options doing validations. This removes the need to keep maps to/from the logger implementation.

Reply:
> the methods which use logrus can be abstracted out and moved to a sub-package to keep it cleaner.

I was thinking that sub-pkg is another over-engineering. Once we switch to something else, we can just remove logrus. no reason to keep logrus in our codebase.

> because you give much more freedom to the actual logger implementation rather than having Options doing validations. 

Yeah, validation for cmd flags can be done in logger via SetOutputLevel, but as I mentioned above, we need to define the supported log-levels for Dapr, not relying on Logrus.ParseLevel. Please wait for my next changes.

Reply:
> I was thinking that sub-pkg is another over-engineering. Once we switch to something else, we can just remove logrus. no reason to keep logrus in our codebase.

Indeed, that's why I mentioned it was not needed. removing `logrus` from `logger.go` will be easy.

> Yeah, validation for cmd flags can be done in logger via SetOutputLevel, but as I mentioned above, we need to define the supported log-levels for Dapr, not relying on Logrus.ParseLevel. Please wait for my next changes.

Again, this does not mean that `ParseLevel` will happen in `options.go`. It means it will happen in `logger.go`, where the other `logrus` specific implementations happen. It will be called from `Options` in `ApplyOptionsToLoggers`.